## Wk3 Proj Submission Toronto Neighborhood Clustering
***************************

### Import the required libraries 

In [1]:
import requests # requests Library
import random # library for random number generation
import pandas as pd # library for data analsysis
import numpy as np # library to handle data in a vectorized manner
from bs4 import BeautifulSoup # Library to handle html

### Read Canada webpage data
Use the Request Library to read Wikipedia site data into a variable

In [3]:
link = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M' #create a link varible containing the html address
page = requests.get(link) # create variable that conatins the returned resulsts from request



Use the Beautiful Soup library to parce webpage text data.
Identitify the desired table and add table data to a Pandas DataFrame. 
<br/>
View DataFrame sample 

In [4]:
page_soup = BeautifulSoup(page.text, 'html.parser') # use Beautiful Soup html parser to handle html code
t = page_soup.find('tbody') # Find First Table find_all tables also an option (Postal codes are included in the first table)
col = t.find_all('th') # Identify the Col Names to create empty Data Frame
rows = t.find_all('tr')[1:] # Identify Row Data to insert into empty Data Frame 

# loop through Html Columns, convert to strings and store in list t1
t1 = []
for i in range(len(col)):
    t1.append(str(col[i]).strip('<th>').strip('</').strip('\n'))

hoods = pd.DataFrame(columns = t1) # create data frame with colums headers using t1 list 

# loop through table rows contained in rows variable and add to the dataframe 
for i in range(len(rows)):
    hoods.loc[i, 'Postcode'] = rows[i].findAll(text=True)[1]
    hoods.loc[i, 'Borough'] = rows[i].findAll(text=True)[3]
    hoods.loc[i, 'Neighbourhood'] = str(rows[i].findAll(text=True)[5]).strip('\n')

print("DataFrame Shape:", hoods.shape, "Rows:", hoods.shape[0], "Columns:", hoods.shape[1]) # inspect dataframe shape
hoods.head() # review dataframe sample

DataFrame Shape: (289, 3) Rows: 289 Columns: 3


Postcode           Borough     Neighbourhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

### Clean and prepare data for clustering 

Drop all rows where Borough is "Not assigned"

In [8]:
df1 = hoods.drop(hoods[hoods.Borough == 'Not assigned'].index) # drop all row's where Borough is equal to "Not assigned" and create a new dataframe
df1 = df1.reset_index(drop=True) # reset new dataframe index  
print("DataFrame Shape:", df1.shape, "Rows:", df1.shape[0], "Columns:", df1.shape[1] ) # inspect new dataframe shape (it should contain less rows)
df1.head() # display a 10 row sample of the new dataframe (Showing no Boroughs = "Not assigned")

DataFrame Shape: (212, 3) Rows: 212 Columns: 3


Postcode           Borough     Neighbourhood
0      M3A        North York         Parkwoods
1      M4A        North York  Victoria Village
2      M5A  Downtown Toronto      Harbourfront
3      M5A  Downtown Toronto       Regent Park
4      M6A        North York  Lawrence Heights

Update all rows where Neighborhood is "Not assigned". Replace "Not assigned" with the Borough name.

In [10]:
# Iterate over the Values in the "Neighborhood" column and update None or Not Assigned values with the Borough name
for i in range(len(df1)):
    if df1['Neighbourhood'][i] == "Not assigned":
        df1['Neighbourhood'][i] = df1['Borough'][i]
    else:
        pass
print("DataFrame Shape:", df1.shape, "Rows:", df1.shape[0], "Columns:", df1.shape[1] ) # inspect dataframe shape (it should not change (212,3))
df1.head() # display a sample of the updated dataframe (all Niehborhoods now showing a value)

DataFrame Shape: (212, 3) Rows: 212 Columns: 3


Postcode           Borough     Neighbourhood
0      M3A        North York         Parkwoods
1      M4A        North York  Victoria Village
2      M5A  Downtown Toronto      Harbourfront
3      M5A  Downtown Toronto       Regent Park
4      M6A        North York  Lawrence Heights

Group data by Post Code.<br/>
And Include each Unique Listed Neighborhood 

In [12]:
boro = {k: g["Borough"].tolist() for k,g in df1.groupby("Postcode")} # Create dictionary with Postal code as the Key, and all corresponding Borough's as a list
hood = {k: g["Neighbourhood"].tolist() for k,g in df1.groupby("Postcode")} # Create a dictionary with Postal code as the Key, and all corresponding Neiborhood's as a list
t2 = pd.DataFrame(list(boro.keys()), columns = ['PostCode'])
for i, row in t2.iterrows():
    t2.loc[i,'Borough'] = list(set(list(boro.values())[i]))[0]    
    t2.loc[i,'Neighbourhood'] = ', '.join(list(set(list(hood.values())[i])))

print("DataFrame Shape:", t2.shape, "Rows:", t2.shape[0], "Columns:", t2.shape[1] ) # inspect dataframe shape (it should not change)
t2.head(10)

DataFrame Shape: (103, 3) Rows: 103 Columns: 3


PostCode      Borough                                    Neighbourhood
0      M1B  Scarborough                                   Malvern, Rouge
1      M1C  Scarborough           Highland Creek, Rouge Hill, Port Union
2      M1E  Scarborough                Morningside, Guildwood, West Hill
3      M1G  Scarborough                                           Woburn
4      M1H  Scarborough                                        Cedarbrae
5      M1J  Scarborough                              Scarborough Village
6      M1K  Scarborough      Ionview, Kennedy Park, East Birchmount Park
7      M1L  Scarborough                  Golden Mile, Clairlea, Oakridge
8      M1M  Scarborough  Cliffside, Scarborough Village West, Cliffcrest
9      M1N  Scarborough                      Cliffside West, Birch Cliff

Check table results for accuracy

In [13]:
t2[(t2['PostCode'] == 'M5A')] # test results by verifying 'M5A' Borough and Neighbourhood

PostCode           Borough              Neighbourhood
53      M5A  Downtown Toronto  Harbourfront, Regent Park

### Adding Longitude and Latitude
--------

Use Requests Library to get Longitude and Latitude information for Postcodes

In [15]:
url="http://cocl.us/Geospatial_data" # create url variable
HoodCords=pd.read_csv(url) # read longitude latidude csv file information with url  
HoodCords.rename(index=str,columns={'Postal Code':'PostCode'},inplace=True) # update Postal Code Column name for dataframe merge 

Merge PostCode DataFrame with the Longitude and Latitude DataFrame 

In [16]:
Hoods_LongLaT = pd.merge(t2, HoodCords, how='left', on='PostCode') # Merge Toronto Neighborhood dataframe (t2) with Long Lat dataframe on Like Postal Codes 
print("DataFrame Shape:", Hoods_LongLaT.shape, "Rows:", Hoods_LongLaT.shape[0], "Columns:", Hoods_LongLaT.shape[1] ) # inspect dataframe shape with 2 new columns (row count should not change)
Hoods_LongLaT.head(15)

DataFrame Shape: (103, 5) Rows: 103 Columns: 5


PostCode      Borough                                      Neighbourhood  \
0       M1B  Scarborough                                     Malvern, Rouge   
1       M1C  Scarborough             Highland Creek, Rouge Hill, Port Union   
2       M1E  Scarborough                  Morningside, Guildwood, West Hill   
3       M1G  Scarborough                                             Woburn   
4       M1H  Scarborough                                          Cedarbrae   
5       M1J  Scarborough                                Scarborough Village   
6       M1K  Scarborough        Ionview, Kennedy Park, East Birchmount Park   
7       M1L  Scarborough                    Golden Mile, Clairlea, Oakridge   
8       M1M  Scarborough    Cliffside, Scarborough Village West, Cliffcrest   
9       M1N  Scarborough                        Cliffside West, Birch Cliff   
10      M1P  Scarborough  Scarborough Town Centre, Wexford Heights, Dors...   
11      M1R  Scarborough                                  Wexford, Maryvale   
12      M1S  Scarborough                                          Agincourt   
13      M1T  Scarborough            Sullivan, Clarks Corners, Tam O'Shanter   
14      M1V  Scarborough  Agincourt North, Steeles East, L'Amoreaux East...   

     Latitude  Longitude  
0   43.806686 -79.194353  
1   43.784535 -79.160497  
2   43.763573 -79.188711  
3   43.770992 -79.216917  
4   43.773136 -79.239476  
5   43.744734 -79.239476  
6   43.727929 -79.262029  
7   43.711112 -79.284577  
8   43.716316 -79.239476  
9   43.692657 -79.264848  
10  43.757410 -79.273304  
11  43.750072 -79.295849  
12  43.794200 -79.262029  
13  43.781638 -79.304302  
14  43.815252 -79.284577

The DataFrame contains 103 rows of information for 11 Unique Boroughs.<br/>
Each borough may contain several different neighborhods, associated to several postal codes.  

In [23]:
print("Unique Borough Count:", len(Hoods_LongLaT['Borough'].unique()))
print("Unique Borough List", Hoods_LongLaT['Borough'].unique())

Unique Borough Count: 11
Unique Borough List ['Scarborough' 'North York' 'East York' 'East Toronto' 'Central Toronto'
 'Downtown Toronto' 'York' 'West Toronto' "Queen's Park" 'Mississauga'
 'Etobicoke']


### Neighborhood Analysis and Clustering
***********

Import Plottnig and Mapping Libraries

In [25]:
import matplotlib.pyplot as plt # plotting library
# backend for rendering plots within the browser
%matplotlib inline 

from sklearn.cluster import KMeans # import k-means from clustering stage
from sklearn.datasets.samples_generator import make_blobs

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors
import json # library to handle JSON files
import folium # folium for map rendering

Display a sample map of Toronto with all PostCode marked and labeled 

In [26]:
lat = 43.70011
long =-79.4163
Toronto_Map = folium.Map(location=[lat, long], zoom_start=11)
# add markers to map
for lat, lng, label, hood in zip(Hoods_LongLaT['Latitude'], Hoods_LongLaT['Longitude'], Hoods_LongLaT['PostCode'], Hoods_LongLaT['Neighbourhood']):
    label = folium.Popup("PostCode:"+label+"\n"+"Neighborhood(s):"+hood, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(Toronto_Map) 

In [27]:
Toronto_Map

Double-click __here__ FourSquare Authentication.
<!--
CLIENT_ID = '' # your Foursquare ID
CLIENT_SECRET = '' # your Foursquare Secret
VERSION = '20190101' # Foursquare API version
-->

Assign Data Return limits and a radius for Four Square

In [30]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius

Define a function that will make each FourSquare API call in relation to the given PostCode and provided Log. & Lat. values. <br/>
This function will print the PostCode and Borough name for each API call.<br/>
It will place the results in a new DataFrame

In [40]:
def getNearbyVenues(names, hoods, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, hood, lat, lng in zip(names, hoods, latitudes, longitudes):
        print(name, hood)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name,
            hood,
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['PostCode', 'Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Use getNearbyVenues function to make FourSquare API calls and store the results.

In [41]:
TorontoVenues = getNearbyVenues(names=Hoods_LongLaT['PostCode'],
                                   hoods=Hoods_LongLaT['Borough'],
                                   latitudes=Hoods_LongLaT['Latitude'],
                                   longitudes=Hoods_LongLaT['Longitude']
                                  )

M1B Scarborough
M1C Scarborough
M1E Scarborough
M1G Scarborough
M1H Scarborough
M1J Scarborough
M1K Scarborough
M1L Scarborough
M1M Scarborough
M1N Scarborough
M1P Scarborough
M1R Scarborough
M1S Scarborough
M1T Scarborough
M1V Scarborough
M1W Scarborough
M1X Scarborough
M2H North York
M2J North York
M2K North York
M2L North York
M2M North York
M2N North York
M2P North York
M2R North York
M3A North York
M3B North York
M3C North York
M3H North York
M3J North York
M3K North York
M3L North York
M3M North York
M3N North York
M4A North York
M4B East York
M4C East York
M4E East Toronto
M4G East York
M4H East York
M4J East York
M4K East Toronto
M4L East Toronto
M4M East Toronto
M4N Central Toronto
M4P Central Toronto
M4R Central Toronto
M4S Central Toronto
M4T Central Toronto
M4V Central Toronto
M4W Downtown Toronto
M4X Downtown Toronto
M4Y Downtown Toronto
M5A Downtown Toronto
M5B Downtown Toronto
M5C Downtown Toronto
M5E Downtown Toronto
M5G Downtown Toronto
M5H Downtown Toronto
M5J Downtow

Inspect the resulting DataFrame by reviewing the shape and a sample of the first 15 rows and the amount of unique Categories.


In [72]:
print('There are {} uniques categories.'.format(len(TorontoVenues['Venue Category'].unique())))
print(TorontoVenues.shape)
TorontoVenues.head(15)

There are 277 uniques categories.
(2233, 8)


PostCode Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0       M1B  Scarborough              43.806686              -79.194353   
1       M1C  Scarborough              43.784535              -79.160497   
2       M1E  Scarborough              43.763573              -79.188711   
3       M1E  Scarborough              43.763573              -79.188711   
4       M1E  Scarborough              43.763573              -79.188711   
5       M1E  Scarborough              43.763573              -79.188711   
6       M1E  Scarborough              43.763573              -79.188711   
7       M1E  Scarborough              43.763573              -79.188711   
8       M1G  Scarborough              43.770992              -79.216917   
9       M1G  Scarborough              43.770992              -79.216917   
10      M1G  Scarborough              43.770992              -79.216917   
11      M1H  Scarborough              43.773136              -79.239476   
12      M1H  Scarborough              43.773136              -79.239476   
13      M1H  Scarborough              43.773136              -79.239476   
14      M1H  Scarborough              43.773136              -79.239476   

                              Venue  Venue Latitude  Venue Longitude  \
0                           Wendy's       43.807448       -79.199056   
1             Royal Canadian Legion       43.782533       -79.163085   
2   Swiss Chalet Rotisserie & Grill       43.767697       -79.189914   
3                 G & G Electronics       43.765309       -79.191537   
4                  Big Bite Burrito       43.766299       -79.190720   
5             Enterprise Rent-A-Car       43.764076       -79.193406   
6             Woburn Medical Centre       43.766631       -79.192286   
7                          Eggsmart       43.767800       -79.190466   
8                         Starbucks       43.770037       -79.221156   
9                       Tim Hortons       43.770827       -79.223078   
10               Korean Grill House       43.770812       -79.214502   
11              Federick Restaurant       43.774697       -79.241142   
12         Drupati's Roti & Doubles       43.775222       -79.241678   
13                      Thai One On       43.774468       -79.241268   
14     Centennial Recreation Centre       43.774593       -79.236500   

          Venue Category  
0   Fast Food Restaurant  
1                    Bar  
2            Pizza Place  
3      Electronics Store  
4     Mexican Restaurant  
5    Rental Car Location  
6         Medical Center  
7         Breakfast Spot  
8            Coffee Shop  
9            Coffee Shop  
10     Korean Restaurant  
11      Hakka Restaurant  
12  Caribbean Restaurant  
13       Thai Restaurant  
14    Athletics & Sports

Use One Hot Encoding to create a numeric value based on each Category for each PostCode  <br/>

In [78]:
# one hot encoding
tv_onehot = pd.get_dummies(TorontoVenues[['Venue Category']], prefix="", prefix_sep="")

# add postcode column back to dataframe
tv_onehot['PostCode'] = TorontoVenues['PostCode'] 

# move postcode column to the first column
fixed_columns = [tv_onehot.columns[-1]] + list(tv_onehot.columns[:-1])
tv_onehot = tv_onehot[fixed_columns]

tv_onehot.head(5)

PostCode  Accessories Store  Adult Boutique  Afghan Restaurant  Airport  \
0      M1B                  0               0                  0        0   
1      M1C                  0               0                  0        0   
2      M1E                  0               0                  0        0   
3      M1E                  0               0                  0        0   
4      M1E                  0               0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  ...  Vegetarian / Vegan Restaurant  Video Game Store  \
0                 0  ...                              0                 0   
1                 0  ...                              0                 0   
2                 0  ...                              0                 0   
3                 0  ...                              0                 0   
4                 0  ...                              0                 0   

   Video Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  \
0            0                      0                0         0          0   
1            0                      0                0         0          0   
2            0                      0                0         0          0   
3            0                      0                0         0          0   
4            0                      0                0         0          0   

   Wings Joint  Women's Store  Yoga Studio  
0            0              0            0  
1            0              0            0  
2            0              0            0  
3            0              0            0  
4            0              0            0  

[5 rows x 278 columns]

Group the One Hot Encoded DataFrame by PostCode for Clustering

In [85]:
print('DataFrame Shape After One Hot encoding:', tv_onehot.shape)

tv_grp = tv_onehot.groupby('PostCode').mean().reset_index() # Group One Hot encoded DataFrame by PostCode

print("DataFrame Shape after grouping by PostCode:", tv_grp.shape)
tv_grp.head()

DataFrame Shape After one Hot encoding: (2233, 278)
DataFrame Shape after grouping by PostCode: (100, 278)


PostCode  Accessories Store  Adult Boutique  Afghan Restaurant  Airport  \
0      M1B                0.0             0.0                0.0      0.0   
1      M1C                0.0             0.0                0.0      0.0   
2      M1E                0.0             0.0                0.0      0.0   
3      M1G                0.0             0.0                0.0      0.0   
4      M1H                0.0             0.0                0.0      0.0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                 0.0           0.0             0.0              0.0   
1                 0.0           0.0             0.0              0.0   
2                 0.0           0.0             0.0              0.0   
3                 0.0           0.0             0.0              0.0   
4                 0.0           0.0             0.0              0.0   

   Airport Terminal  ...  Vegetarian / Vegan Restaurant  Video Game Store  \
0               0.0  ...                            0.0               0.0   
1               0.0  ...                            0.0               0.0   
2               0.0  ...                            0.0               0.0   
3               0.0  ...                            0.0               0.0   
4               0.0  ...                            0.0               0.0   

   Video Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  \
0          0.0                    0.0              0.0       0.0        0.0   
1          0.0                    0.0              0.0       0.0        0.0   
2          0.0                    0.0              0.0       0.0        0.0   
3          0.0                    0.0              0.0       0.0        0.0   
4          0.0                    0.0              0.0       0.0        0.0   

   Wings Joint  Women's Store  Yoga Studio  
0          0.0            0.0          0.0  
1          0.0            0.0          0.0  
2          0.0            0.0          0.0  
3          0.0            0.0          0.0  
4          0.0            0.0          0.0  

[5 rows x 278 columns]

Define a function that takes the one hot encoded values and identifies the 10 most common venues for each PostCode. <br/>
Visualizing this data will help to review the Common Venue rankings when comparing PostCodes that have been clustered together.

In [87]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['PostCode']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
pc_venues_sorted = pd.DataFrame(columns=columns)
pc_venues_sorted['PostCode'] = tv_grp['PostCode']

for ind in np.arange(tv_grp.shape[0]):
    pc_venues_sorted.iloc[ind, 1:] = return_most_common_venues(tv_grp.iloc[ind, :], num_top_venues)

pc_venues_sorted.head()

PostCode 1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0      M1B  Fast Food Restaurant            Donut Shop          Dessert Shop   
1      M1C                   Bar           Yoga Studio             Drugstore   
2      M1E    Mexican Restaurant           Pizza Place     Electronics Store   
3      M1G           Coffee Shop     Korean Restaurant             Drugstore   
4      M1H                Lounge      Hakka Restaurant                Bakery   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0    Dim Sum Restaurant                 Diner        Discount Store   
1                 Diner        Discount Store               Dog Run   
2        Breakfast Spot        Medical Center   Rental Car Location   
3    Dim Sum Restaurant                 Diner        Discount Store   
4    Athletics & Sports                  Bank       Thai Restaurant   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0               Dog Run      Doner Restaurant             Drugstore   
1      Doner Restaurant            Donut Shop   Dumpling Restaurant   
2           Yoga Studio    Dim Sum Restaurant                 Diner   
3               Dog Run      Doner Restaurant            Donut Shop   
4  Caribbean Restaurant   Fried Chicken Joint           Yoga Studio   

  10th Most Common Venue  
0         Farmers Market  
1           Dessert Shop  
2         Discount Store  
3            Yoga Studio  
4       Doner Restaurant

Run the Kmeans Clustering alogrithim in an attempt to Cluster PostCodes <br/>
Print a sample of the Cluster Labels

In [89]:
# set number of clusters
kclusters = 5
torontoGrpClust = tv_grp.drop('PostCode', 1)

kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(torontoGrpClust) # run k-means clustering
kmeans.labels_[0:10] # check cluster labels generated for each row in the dataframe

array([0, 2, 2, 2, 2, 0, 2, 2, 2, 2], dtype=int32)

Insert the new Cluster Labels clumns into the sorted venues datatframe <br/>
Then merge sorted venues DataFrame with the Most Common Venues Data Frame 

In [90]:
# add clustering labels
pc_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

tMerged = Hoods_LongLaT
tMerged = tMerged.join(pc_venues_sorted.set_index('PostCode'), on='PostCode') # merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood

tMerged.head() # check the last columns!

PostCode      Borough                           Neighbourhood   Latitude  \
0      M1B  Scarborough                          Malvern, Rouge  43.806686   
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2      M1E  Scarborough       Morningside, Guildwood, West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.194353             0.0  Fast Food Restaurant            Donut Shop   
1 -79.160497             2.0                   Bar           Yoga Studio   
2 -79.188711             2.0    Mexican Restaurant           Pizza Place   
3 -79.216917             2.0           Coffee Shop     Korean Restaurant   
4 -79.239476             2.0                Lounge      Hakka Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0          Dessert Shop    Dim Sum Restaurant                 Diner   
1             Drugstore                 Diner        Discount Store   
2     Electronics Store        Breakfast Spot        Medical Center   
3             Drugstore    Dim Sum Restaurant                 Diner   
4                Bakery    Athletics & Sports                  Bank   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0        Discount Store               Dog Run      Doner Restaurant   
1               Dog Run      Doner Restaurant            Donut Shop   
2   Rental Car Location           Yoga Studio    Dim Sum Restaurant   
3        Discount Store               Dog Run      Doner Restaurant   
4       Thai Restaurant  Caribbean Restaurant   Fried Chicken Joint   

  9th Most Common Venue 10th Most Common Venue  
0             Drugstore         Farmers Market  
1   Dumpling Restaurant           Dessert Shop  
2                 Diner         Discount Store  
3            Donut Shop            Yoga Studio  
4           Yoga Studio       Doner Restaurant

In [93]:
tMerged['Cluster Labels'].fillna(0, inplace=True)

Use Folium to plot a new Map of Toronto PostCodes where Point Color corresponds to Assigned Cluster Value 

In [95]:
# create map
Toronto_clusters = folium.Map(location=[lat, long], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(tMerged['Latitude'], tMerged['Longitude'], tMerged['PostCode'], tMerged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster)-1],
        fill=True,
        fill_color=rainbow[int(round(cluster))-1],
        fill_opacity=0.7).add_to(Toronto_clusters)    
Toronto_clusters

### Inspecting Cluster results
Print each PostCode that was assigned to the selected cluster.
* Cluster 0: Has a High Frequency of Coffee Shops in the top results
* Cluster 1: Has 2 Post codes and The Most common venues for those are a Baseball field, and a yoga studio. These two track very closely in all top venues.
* Cluster 2: Have Parks, Yoga Studios and a dumpling Restaurant in the top venues 
* Cluster 3: Has the Fewest with only 1 PostCode where  the Bank, Eastern European Restaurants and Discount Stores are the top venues  
* Cluster 4: Has the most Post codes and seems to have the Highest frequency of Restaurants. As the most Common Venues

In [73]:
tMerged.loc[tMerged['Cluster Labels'] == 0, tMerged.columns[[0] + list(range(5, tMerged.shape[1]))]]

PostCode  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
3       M1G             0.0           Coffee Shop     Korean Restaurant   
6       M1K             0.0      Department Store         Train Station   
16      M1X             0.0                   NaN                   NaN   
24      M2R             0.0           Pizza Place         Grocery Store   
29      M3J             0.0                   Bar        Massage Studio   
34      M4A             0.0            Nail Salon         Grocery Store   
37      M4E             0.0                   Pub     Health Food Store   
48      M4T             0.0                   NaN                   NaN   
49      M4V             0.0                   Pub           Coffee Shop   
53      M5A             0.0           Coffee Shop                Bakery   
57      M5G             0.0           Coffee Shop                  Café   
59      M5J             0.0           Coffee Shop                 Hotel   
60      M5K             0.0           Coffee Shop                  Café   
80      M6M             0.0    Turkish Restaurant           Coffee Shop   
85      M7A             0.0           Coffee Shop                   Gym   
86      M7R             0.0           Coffee Shop                 Hotel   
93      M9A             0.0                   NaN                   NaN   
98      M9N             0.0                   NaN                   NaN   

   3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
3            Yoga Studio        Discount Store                      Dog Run   
6         Discount Store           Coffee Shop           Chinese Restaurant   
16                   NaN                   NaN                          NaN   
24           Coffee Shop               Butcher                     Pharmacy   
29           Coffee Shop  Caribbean Restaurant  Eastern European Restaurant   
34           Coffee Shop          Hockey Arena        Portuguese Restaurant   
37           Coffee Shop          Neighborhood                  Yoga Studio   
48                   NaN                   NaN                          NaN   
49           Supermarket    Light Rail Station        Vietnamese Restaurant   
53                  Park                  Café                          Pub   
57    Italian Restaurant                   Bar                 Burger Joint   
59              Aquarium           Pizza Place                         Café   
60                 Hotel            Restaurant          American Restaurant   
80        Sandwich Place        Discount Store                    Drugstore   
85   Japanese Restaurant                 Diner             Sushi Restaurant   
86   American Restaurant   Fried Chicken Joint     Mediterranean Restaurant   
93                   NaN                   NaN                          NaN   
98                   NaN                   NaN                          NaN   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
3       Doner Restaurant            Donut Shop             Drugstore   
6                  Diner               Dog Run      Doner Restaurant   
16                   NaN                   NaN                   NaN   
24      Greek Restaurant  Ethiopian Restaurant     Electronics Store   
29               Dog Run      Doner Restaurant            Donut Shop   
34           Yoga Studio             Drugstore                 Diner   
37                 Diner        Discount Store               Dog Run   
48                   NaN                   NaN                   NaN   
49   Fried Chicken Joint           Pizza Place            Bagel Shop   
53        Breakfast Spot    Mexican Restaurant               Theater   
57        Sandwich Place       Thai Restaurant        Ice Cream Shop   
59            Restaurant        Scenic Lookout                Bakery   
60             Gastropub         Deli / Bodega    Italian Restaurant   
80                 Diner               Dog Run      Doner Restauran

In [74]:
tMerged.loc[tMerged['Cluster Labels'] == 1, tMerged.columns[[0] + list(range(5, tMerged.shape[1]))]]

PostCode  Cluster Labels       1st Most Common Venue 2nd Most Common Venue  \
91      M8Y             1.0              Baseball Field           Yoga Studio   
97      M9M             1.0  Construction & Landscaping        Baseball Field   

          3rd Most Common Venue        4th Most Common Venue  \
91  Eastern European Restaurant               Discount Store   
97                  Yoga Studio  Eastern European Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
91               Dog Run      Doner Restaurant            Donut Shop   
97               Dog Run      Doner Restaurant            Donut Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
91             Drugstore   Dumpling Restaurant      Electronics Store  
97             Drugstore   Dumpling Restaurant    Empanada Restaurant

In [75]:
tMerged.loc[tMerged['Cluster Labels'] == 2, tMerged.columns[[0] + list(range(5, tMerged.shape[1]))]]

PostCode  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
14       M1V             2.0            Playground           Coffee Shop   
20       M2L             2.0                  Park           Yoga Studio   
23       M2P             2.0                   Bar                  Bank   
30       M3K             2.0                  Park               Airport   
40       M4J             2.0                  Park           Coffee Shop   
50       M4W             2.0                  Park                 Trail   
74       M6E             2.0                  Park              Pharmacy   
90       M8X             2.0                 River                  Park   
100      M9R             2.0           Pizza Place                  Park   

           3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
14                          Park           Yoga Studio             Drugstore   
20   Eastern European Restaurant        Discount Store               Dog Run   
23                          Park               Dog Run      Doner Restaurant   
30                      Bus Stop           Yoga Studio   Dumpling Restaurant   
40             Convenience Store           Yoga Studio   Dumpling Restaurant   
50                    Playground           Yoga Studio    Dim Sum Restaurant   
74                        Market  Fast Food Restaurant         Women's Store   
90                   Yoga Studio            Donut Shop    Dim Sum Restaurant   
100          Dumpling Restaurant                 Diner        Discount Store   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
14                  Diner        Discount Store               Dog Run   
20       Doner Restaurant            Donut Shop             Drugstore   
23             Donut Shop             Drugstore   Dumpling Restaurant   
30         Discount Store               Dog Run      Doner Restaurant   
40         Discount Store               Dog Run      Doner Restaurant   
50                  Diner        Discount Store               Dog Run   
74          Grocery Store      Department Store  Ethiopian Restaurant   
90                  Diner        Discount Store               Dog Run   
100               Dog Run      Doner Restaurant            Donut Shop   

           9th Most Common Venue       10th Most Common Venue  
14              Doner Restaurant                   Donut Shop  
20           Dumpling Restaurant            Electronics Store  
23   Eastern European Restaurant                  Yoga Studio  
30                    Donut Shop                    Drugstore  
40                    Donut Shop                    Drugstore  
50              Doner Restaurant                   Donut Shop  
74           Empanada Restaurant            Electronics Store  
90              Doner Restaurant                    Drugstore  
100                    Drugstore  Eastern European Restaurant

In [76]:
tMerged.loc[tMerged['Cluster Labels'] == 3, tMerged.columns[[0] + list(range(5, tMerged.shape[1]))]]

PostCode  Cluster Labels 1st Most Common Venue  \
94      M9B             3.0                  Bank   

          2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
94  Eastern European Restaurant        Discount Store               Dog Run   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
94      Doner Restaurant            Donut Shop             Drugstore   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
94   Dumpling Restaurant           Yoga Studio                  Field

In [77]:
tMerged.loc[tMerged['Cluster Labels'] == 4, tMerged.columns[[0] + list(range(5, tMerged.shape[1]))]]

PostCode  Cluster Labels    1st Most Common Venue  \
0        M1B             4.0     Fast Food Restaurant   
1        M1C             4.0                      Bar   
2        M1E             4.0           Breakfast Spot   
4        M1H             4.0         Hakka Restaurant   
5        M1J             4.0              Pizza Place   
7        M1L             4.0                   Bakery   
8        M1M             4.0                    Motel   
9        M1N             4.0          College Stadium   
10       M1P             4.0        Indian Restaurant   
11       M1R             4.0              Auto Garage   
12       M1S             4.0           Breakfast Spot   
13       M1T             4.0              Pizza Place   
15       M1W             4.0       Chinese Restaurant   
17       M2H             4.0                  Dog Run   
18       M2J             4.0           Clothing Store   
19       M2K             4.0                     Bank   
21       M2M             4.0                      Gym   
22       M2N             4.0         Ramen Restaurant   
25       M3A             4.0        Food & Drink Shop   
26       M3B             4.0           Baseball Field   
27       M3C             4.0              Coffee Shop   
28       M3H             4.0              Coffee Shop   
31       M3L             4.0            Grocery Store   
32       M3M             4.0         Business Service   
33       M3N             4.0     Gym / Fitness Center   
35       M4B             4.0     Fast Food Restaurant   
36       M4C             4.0             Skating Rink   
38       M4G             4.0              Coffee Shop   
39       M4H             4.0        Indian Restaurant   
41       M4K             4.0         Greek Restaurant   
..       ...             ...                      ...   
62       M5M             4.0              Coffee Shop   
63       M5N             4.0  Health & Beauty Service   
64       M5P             4.0                 Bus Line   
65       M5R             4.0              Coffee Shop   
66       M5S             4.0                     Café   
67       M5T             4.0                     Café   
68       M5V             4.0         Airport Terminal   
69       M5W             4.0              Coffee Shop   
70       M5X             4.0              Coffee Shop   
71       M6A             4.0           Clothing Store   
72       M6B             4.0                      Pub   
73       M6C             4.0                    Field   
75       M6G             4.0                     Café   
76       M6H             4.0                   Bakery   
77       M6J             4.0                      Bar   
78       M6K             4.0              Coffee Shop   
79       M6L             4.0                     Park   
81       M6N             4.0              Pizza Place   
82       M6P             4.0                     Café   
83       M6R             4.0           Breakfast Spot   
84       M6S             4.0              Coffee Shop   
87       M7Y             4.0       Light Rail Station   
88       M8V             4.0                 Pharmacy   
89       M8W             4.0              Pizza Place   
92       M8Z             4.0            Burrito Place   
95       M9C             4.0                 Pharmacy   
96       M9L             4.0              Pizza Place   
99       M9P             4.0              Pizza Place   
101      M9V             4.0            Grocery Store   
102      M9W             4.0                Drugstore   

          2nd Most Common Venue        3rd Most Common Venue  \
0                    Print Shop                    Drugstore   
1                   Golf Course  Eastern European Restaurant   
2                           Spa                  Pizza Place   
4                          Bank                       Lounge   
5                    Playground                    Drugstore   
7                      Bus Line         Fast Food Restaurant   
8           American 

### After reviewing Cluster results edit Kmeans calculation and re-run to review result differences <br/>
Increased from 5 Clusters to 8 to see if the data would be grouped in a similar fashion
Plot results with Toronto Clusters 2 Map

Additional Results review and comparison can be done by displaying the PostCode by Cluster Group as done previsouly 


In [62]:
# set number of clusters
kclusters2 = 8

torontoGrpClust2 = tv_grp.drop('PostCode', 1)

# run k-means clustering
kmeans2 = KMeans(n_clusters=kclusters2, random_state=0).fit(torontoGrpClust2)

# check cluster labels generated for each row in the dataframe
kmeans2.labels_

array([1, 4, 2, 2, 2, 6, 2, 0, 2, 2, 2, 2, 2, 2, 6, 2, 2, 2, 2, 3, 2, 7,
       2, 0, 2, 2, 2, 2, 0, 2, 2, 2, 2, 2, 2, 2, 2, 2, 0, 2, 2, 2, 0, 2,
       2, 2, 6, 2, 6, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 0, 0, 0, 2, 2, 2, 2, 0, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       0, 2, 2, 7, 2, 2, 5, 2, 2, 0, 2, 2], dtype=int32)

In [63]:
# add clustering labels
pcs2.insert(0, 'Cluster Labels', kmeans2.labels_)

tMerged2 = Hoods_LongLaT

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
tMerged2 = tMerged2.join(pcs2.set_index('PostCode'), on='PostCode')

tMerged2.head() # check the last columns!

PostCode      Borough                           Neighbourhood   Latitude  \
0      M1B  Scarborough                          Malvern, Rouge  43.806686   
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2      M1E  Scarborough       Morningside, Guildwood, West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.194353             1.0  Fast Food Restaurant            Donut Shop   
1 -79.160497             4.0                   Bar           Yoga Studio   
2 -79.188711             2.0    Mexican Restaurant           Pizza Place   
3 -79.216917             2.0           Coffee Shop     Korean Restaurant   
4 -79.239476             2.0                Lounge      Hakka Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0          Dessert Shop    Dim Sum Restaurant                 Diner   
1             Drugstore                 Diner        Discount Store   
2     Electronics Store        Breakfast Spot        Medical Center   
3             Drugstore    Dim Sum Restaurant                 Diner   
4                Bakery    Athletics & Sports                  Bank   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0        Discount Store               Dog Run      Doner Restaurant   
1               Dog Run      Doner Restaurant            Donut Shop   
2   Rental Car Location           Yoga Studio    Dim Sum Restaurant   
3        Discount Store               Dog Run      Doner Restaurant   
4       Thai Restaurant  Caribbean Restaurant   Fried Chicken Joint   

  9th Most Common Venue 10th Most Common Venue  
0             Drugstore         Farmers Market  
1   Dumpling Restaurant           Dessert Shop  
2                 Diner         Discount Store  
3            Donut Shop            Yoga Studio  
4           Yoga Studio       Doner Restaurant

In [64]:
tMerged2['Cluster Labels'].fillna(0, inplace=True)

In [65]:
# create map
Toronto_clusters2 = folium.Map(location=[lat, long], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters2)
ys = [i + x + (i*x)**2 for i in range(kclusters2)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(tMerged2['Latitude'], tMerged2['Longitude'], tMerged2['PostCode'], tMerged2['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster)-1],
        fill=True,
        fill_color=rainbow[int(cluster)-1],
        fill_opacity=0.7).add_to(Toronto_clusters2)
       
Toronto_clusters2